File: 72-edit-distance/edit-distance.java
class Solution {
    public int minDistance(String word1, String word2) {
        int[][] dp=new int[word1.length()][word2.length()];
        for(int[] a:dp){
            Arrays.fill(a,-1);
        }
        return dis(word1,word2,0,0,dp);
    }
    public int dis(String w1,String w2,int i,int j,int[][]dp){
        if(i==w1.length()){
            return w2.length()-j;
        }
        if(j==w2.length()){
            return w1.length()-i;
        }
        if(dp[i][j]!=-1){
            return dp[i][j];
        }
        int ans=0;
        if(w1.charAt(i)==w2.charAt(j)){
            ans=dis(w1,w2,i+1,j+1,dp);
        }
        else{
            int I=dis(w1,w2,i,j+1,dp);
            int D=dis(w1,w2,i+1,j,dp);
            int R=dis(w1,w2,i+1,j+1,dp);
            ans=Math.min(I,Math.min(R,D))+1;
        }
        return dp[i][j]=ans;
    }
}

File: 72-edit-distance/edit-distance.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        vector<vector<int>> dp(word1.size(),vector<int>(word2.size(),-1));
        return count(word1,word2,0,0,dp);
    }
    int count(string &s,string &t,int i,int j,vector<vector<int>> &dp){
        if(i==s.length()){
            return t.length()-j;
        }
        if(j==t.length()){
            return s.length()-i;
        }
        if(dp[i][j]!=-1){
            return dp[i][j];
        }
        int ans=0;
        if(s[i]==t[j]){
            ans=count(s,t,i+1,j+1,dp);
        }
        else{
            int ins=count(s,t,i,j+1,dp);
            int d=count(s,t,i+1,j,dp);
            int r=count(s,t,i+1,j+1,dp);
            ans=min(d,min(r,ins))+1;
        }
        return dp[i][j]=ans;
    }
};